src/_gettsim/time_conversion.py

```diff
@@ -0,0 +1,412 @@
+    -------
+    float
+        Monthly value.
+    """
+    return value * _W_PER_Y / _M_PER_Y
+
+
+def w_to_d(value: float) -> float:
+    """
+    Converts weekly to daily values.
+
+    Parameters
+    ----------
+    value
+        Weekly value to be converted to daily value.
+
+    Returns
+    -------
+    float
+        Daily value.
+    """
+    return value * _W_PER_Y / _D_PER_Y
+
+
+def d_to_y(value: float) -> float:
+    """
+    Converts daily to yearly values.
+
+    Parameters
+    ----------
+    value
+        Daily value to be converted to yearly value.
+
+    Returns
+    -------
+    float
+        Yearly value.
+    """
+    return value * _D_PER_Y
+
+
+def d_to_m(value: float) -> float:
+    """
+    Converts daily to monthly values.
+
+    Parameters
+    ----------
+    value
+        Daily value to be converted to monthly value.
+
+    Returns
+    -------
+    float
+        Monthly value.
+    """
+    return value * _D_PER_Y / _M_PER_Y
+
+
+def d_to_w(value: float) -> float:
+    """
+    Converts daily to weekly values.
+
+    Parameters
+    ----------
+    value
+        Daily value to be converted to weekly value.
+
+    Returns
+    -------
+    float
+        Weekly value.
+    """
+    return value * _D_PER_Y / _W_PER_Y
+
+
+def create_functions_for_time_units(
+    functions: dict[str, Callable],
+    data_cols: list[str],
+) -> dict[str, Callable]:
+    """
+    Create functions for other time units.
+
+    The time unit of a function is determined by a naming convention:
+    * Yearly functions end with "_y", "_y_hh" or "_y_tu".
+    * Monthly functions end with "_m", "_m_hh" or "_m_tu".
+    * Weekly functions end with "_w", "_w_hh" or "_w_tu".
+    * Daily functions end with "_d", "_d_hh" or "_d_tu".
+
+    Unless the corresponding function already exists, the following functions are
+    created:
+    * For yearly functions, create monthly, weekly and daily functions.
+    * For monthly functions, create yearly, weekly and daily functions.
+    * For weekly functions, create yearly, monthly and daily functions.
+    * For daily functions, create yearly, monthly and weekly functions.
+
+    Parameters
+    ----------
+    functions
+        Dictionary of functions.
+    data_cols
+        List of data columns.
+
+    Returns
+    -------
+    dict[str, Callable]
+        Dictionary of created functions.
+    """
+
+    result = {}
+
+    for name, func in functions.items():
+        result.update(_create_functions_for_time_units(name, func))
+
+    for name in data_cols:
+        result.update(_create_functions_for_time_units(name))
+
+    return {
+        name: func
+        for name, func in result.items()
+        if name not in functions and name not in data_cols
+    }
+
+
+def _create_functions_for_time_units(  # noqa: PLR0915
+    name: str, func: Callable | None = None
+) -> dict[str, Callable]:
+    result = {}
+    info = getattr(func, "__info__", None)
+
+    if name.endswith("_y"):
```

Comment:
This block seems very verbose to me. Can't we do the check for the name upfront and then reduce the amount of repetition?

Something like (untested):

```py
time_units = ["y", "m", "w", "d"] # should be some global?
id_groupings = ("", "_hh", "_tu")  # Will change in near future, should be some global?
possible_suffixes = {f"_{unit}{id_g}": f"{unit}" for id_g in id_groupings for unit in time_units}
orig_suffix, orig_unit = [(k, v) for k, v in possible_suffixes.items() if name.endswith(k)][0]
_target_units = [u for u in time_units if not u == orig_unit]
target_suffixes = {orig_suffix.replace(orig_unit, u, 1): u for u in _target_units}
result = {
    _replace_suffix(name, orig_suffix, target_suffix): _create_function_for_time_unit(
        name, info, converter_funcs[f"{orig_unit}_to_{target_unit}"]
    )
    for target_suffix, target_unit in target_suffixes.items()
}
```

where `converter_funcs` is a dict with keys containing string version of the function name. Maybe there is a more elegant solution.